zio-cache/shared/src/main/scala/zio/cache/ManagedCache.scala

```diff
@@ -0,0 +1,388 @@
+package zio.cache
+
+import zio.ZManaged.Finalizer
+import zio._
+import zio.internal.MutableConcurrentQueue
+
+import java.time.{Clock, Duration, Instant}
+import java.util
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger, LongAdder}
+import scala.jdk.CollectionConverters._
+
+sealed abstract class ManagedCache[-Key, +Error, +Value] {
+
+  /**
+   * Returns statistics for this cache.
+   * @return
+   */
+  def cacheStats: UIO[CacheStats]
+
+  /**
+   * Return whether a Managed associated with the specified key exists in the cache.
+   * @param key
+   * @return
+   */
+  def contains(key: Key): UIO[Boolean]
+
+  /**
+   * Return statistics for the specified entry.
+   */
+  def entryStats(key: Key): UIO[Option[EntryStats]]
+
+  /**
+   * Give a proxy managed on the resource already created for this key if any.
```

Comment:
I would more conceptualize this as "gets the value from the cache if it exists or otherwise computes it, the release action signals to the cache that the value is no longer being used and can potentially be finalized subject to the policies of the cache".

Reply:
Went for:
```
  /**
   * Return a managed that gets the value from the cache if it exists or otherwise computes it, the release action
   * signals to the cache that the value is no longer being used and can potentially be finalized subject to the
   * policies of the cache
   * @param key
   * @return
   */
```
Thought `return a managed` was important, because I want to make sure user understand all of that is happening only once he use the returned Managed

Reply:
Okay but that is true for every value in `ZIO`. Regardless of whether it is a `ZManaged` or a `ZIO` nothing happens unless it is evaluated. So at least to me it is confusing in that it implies that something different is going on here whereas what is going on is exactly the same as everything else in ZIO. One thing we should make sure of here is that it should not even look up the value in the cache until the `ZManaged` is evaluated since otherwise that would be a violation of referential transparency.

Reply:
> Okay but that is true for every value in ZIO. Regardless of whether it is a ZManaged or a ZIO nothing happens unless it is evaluated.

That is a really good point, and it make me reconsider the change I did to your proposal. In fact, specifying here the normal behaviour of ZIO is confusing, I went back to your exact proposition.

> One thing we should make sure of here is that it should not even look up the value in the cache until the ZManaged is evaluated since otherwise that would be a violation of referential transparency.
I already had test about making sure that the value isn't lookup before the ZManaged is evaluated! But I didn't about checking that contains is returning false before the ZManaged is evaluated, I just added one